scripts/Vulkan.py

```diff
@@ -43,7 +44,7 @@ def CheckVulkanSDK():
     print(f"Correct Vulkan SDK located at {VULKAN_SDK}")
     return True
 
-VulkanSDKDebugLibsURL = 'https://files.lunarg.com/SDK-1.2.170.0/VulkanSDK-1.2.170.0-DebugLibs.zip'
+VulkanSDKDebugLibsURL = f"https://sdk.lunarg.com/sdk/download/{HAZEL_VULKAN_VERSION}/windows/VulkanSDK-{HAZEL_VULKAN_VERSION}-DebugLibs.zip"
```

Comment:
403 means access denied, could be because the header was not set up correctly? If it was an invalid url, it would throw a 404...

Reply:
Well, I don't know Python at all, but that link surely doesn't work, whereas the original link does.

Reply:
I just tried this link, and it downloads it just fine:
![image](https://user-images.githubusercontent.com/26593477/119549955-116d8180-bd98-11eb-9c2d-f2c11fe29377.png)

Reply:
Is it like, am I missing something which is not present in the script either?

Reply:
![Screenshot_1](https://user-images.githubusercontent.com/26383905/119599479-b58c1280-be02-11eb-9aed-d4ba41432860.png)

This is the error that turns up. NOTE:- I am running this on Ghost Engine, trying not to harm the original commit of Hazel.

Reply:
Downloading with a script is always 'dangerous' for a server as they can see it as an attack on the server. For that reason you have to make sure you're not a 'bot' by providing the browser credentials as you 'click' the download request. By requiring you to have an up-to-date browser, this means that they can block 'old' attacks. This is why you get blocked by a 403 error using the script (as I said earlier).

In python there is a package named UserAgent that is basically a (more or less up to date) library containing the most recent browser credentials which you provide in the header of your request. In hazel the header is created here:
https://github.com/TheCherno/Hazel/blob/93eef739adc2cc862a20e66dc2aad8739502d585/scripts/Utils.py#L10

You can try to:
1. Check if your UserAgent package is up to date (depending on how you installed python in the first place, you can use pip or any other package browser delivered with your installation (I personally prefer Anaconda).
2. If the package is up to date, you could still try to hardcode a header instead. On my own engine I currently use
```py
headers = {'User-Agent': "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36"}
```

I hope this helps you understand more of the reason you're getting a 403 error and what the python script actually does. As I said earlier, if the link was invalid, you would get a 404 error, so the link is definitely working, it's the server that is refusing you to connect.

Reply:
Okay, I changed the headers, as well. Still getting rejected. I believe that, maybe Cherno tried doing this as well, and since this error was turning up, he changed the url to what he put in the original script.

In `Utils.py` I changed the headers to this (Also updated UserAgent package) -
```
headers = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11',
         'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
         'Connection': 'keep-alive'}
         
        response =  requests.get(url, headers=headers, stream=True)
        total = response.headers.get('content-length')
```

Any suggestions regarding this?

Reply:
for me, my download function looks like:
```py
def DownloadFile(url, filepath):
    path = filepath
    filepath = os.path.abspath(filepath)
    try:
        os.makedirs(os.path.dirname(filepath))
    except FileExistsError:
        # Directory already exist
        if os.path.isfile(filepath):
            print(f"Skipping download, {path} already exist!")
            return

    with open(filepath, 'wb') as f:
        headers = {'User-Agent': "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36"}
        response = requests.get(url, headers=headers, stream=True)
        total = response.headers.get('content-length')

        if total is None:
            f.write(response.content)
        else:
            downloaded = 0
            total = int(total)
            startTime = time.time()
            for data in response.iter_content(chunk_size=max(int(total/1000), 1024*1024)):
                downloaded += len(data)
                f.write(data)
                done = int(50*downloaded/total)
                percentage = (downloaded / total) * 100
                elapsedTime = time.time() - startTime
                avgKBPerSecond = (downloaded / 1024) / elapsedTime
                avgSpeedString = '{:.2f} KB/s'.format(avgKBPerSecond)
                if (avgKBPerSecond > 1024):
                    avgMBPerSecond = avgKBPerSecond / 1024
                    avgSpeedString = '{:.2f} MB/s'.format(avgMBPerSecond)
                sys.stdout.write('\r[{}{}] {:.2f}% ({})     '.format('█' * done, '.' * (50-done), percentage, avgSpeedString))
                sys.stdout.flush()
    sys.stdout.write('\n')
```

Reply:
Nope I still get the 403 forbidden error. After exactly setting up stuff like you mentioned.


Along with this there are scenarios which we need to look at :-
* The person who's cloning the repository may or may not have python installed.
* He/she may or may not have an updated version of UserAgent package installed.
* If the headers need to be changed from person to person, that's a big hassle nobody would like to go through.

I am pushing a commit that will change the URL back to what Cherno has set. In the meantime, if you come up with a sure-shot way of ensuring the 403 error doesn't turn up, put it up here. I can try it out on my engine, to see whether it works or not.

Reply:
> I am pushing a commit that will change the URL back to what Cherno has set.

Same issue, as I'm getting a 403 with the commit you've made (that link doesn't work for me)...

Reply:
Changed the URL, there was a typo error, can you try it out?

I also tried your link in Hazel, trying to clear the benefit of doubt that my Engine may be having certain discrepancies, I still get the 403 in Hazel as well. Whereas the Cherno's link works. In both my Engine and Hazel.

Reply:
I tried it with your change and I also tried it again with how Hazel defines it now, but I get the same result as mentioned before in https://github.com/TheCherno/Hazel/pull/438#issuecomment-830766448 ... still a 403

Reply:
So let's do one thing. For now, let's keep the link what it was for the time being. If more people report about it not working on their machine, then we can change it to your version. I believe this would require feedback from more people than us two.

Reply:
Maybe try getting back to Cherno about this one as well.

Reply:
I did not get a response so far, I asked a while back...

Reply:
I can confirm the link I provided is the one mentioned in Hazel-dev at this point in time.

[`https://sdk.lunarg.com/sdk/download/1.2.170.0/windows/VulkanSDK-1.2.170.0-DebugLibs.zip`](https://sdk.lunarg.com/sdk/download/1.2.170.0/windows/VulkanSDK-1.2.170.0-DebugLibs.zip) in particular for the debug libs

Reply:
Can we do something like a try-catch statement, where if the first link fails, then the second link is used? Kinda like a fallback URL?

Reply:
yes we can, should we directly convert this PR to my rewrite, then we can easily modify it to accept a list of urls, and it just tries all of them until one succeeds...

Reply:
Here is my version of my repo, it's a folder with all the python files, which resides inside the scripts folder (to not obfuscate the bat files themselves).
[Python.zip](https://github.com/TheCherno/Hazel/files/6610163/Python.zip)

The thing we should have to change is the `DownloadFile(url, filepath)` function to also accept a list of urls.

Reply:
I tried your scripts from that zip folder. I am getting an error
```
AttributeError: module 'importlib' has no attribute 'util'
```

I have Python 3.9 installed, and a quick Google search reports this back - https://bugs.python.org/issue41958

The above mentioned features would be great. Let's see if we can incorporate all of these features. One thing that is a must-have is the list of URLs, which it tries all of them until one succeeds.

Reply:
according to that bug report, you should import util directly...

I assume you could do it as `import importlib.util as importlib_util` and then use it at line 28 as `if importlib_util.find_spec(packageName) is None:`

can you confirm that this works? I only have 3.8 on my system, so I cannot test this.

Reply:
Okay, so it's downloading, the URL works. But the path strings:
```
vulkanDirectory = "./../Hazel/vendor/VulkanSDK"
```
and
```
subprocess.call([os.path.abspath("./scripts/Win-GenProjects.bat"), "nopause"])
```
aren't working properly. The first string downloads in some random place, and not the intended location that I want. The second command throws a FileNotFoundError [winerror 2]. I believe these are the last two pieces of the puzzle that can fix most issues.

Reply:
In my egine it is in another sub folder...
In hazel, you call it from `scripts/Setup.bat` where the setup.py resides in `scripts/python/Setup.py`

In Setup.py we change from the script folder to the root directory for the Vulkan install, so change line 12 to `os.chdir('./../') # Change from scripts directory to root`  This should also fix the call to the `Win-GenProjects.bat`.

And in SetupVulkan.py you might want to change line 13 `vulkanDirectory` to the correct Hazel location.

While there, line 66 still has a hardcoded version, change that to `vulkanPath = f"{cls.vulkanDirectory}/VulkanSDK-{cls.requiredVulkanVersion}-DebugLibs.zip"`

That should make sure it is in the correct location?

Reply:
Implemented the changes. Working on my end. Please be sure to test it out once on your side as well.

Reply:
I tested it out with python 3.9 and 3.8, both worked for me.
![image](https://user-images.githubusercontent.com/26593477/121557111-c2208580-ca14-11eb-90a8-a720f578b860.png)

The remaining step is to modify the `DownloadFile(url, filepath)` to both work with url being a string, or being a list of strings. This should be fairly straight forward...

```py
def DownloadFile(url, filepath):
    if type(url) == 'list':
        for url_option in url:
            try:
                DownloadFile(url_option, filepath)
                return # return if succeeded
            except:
                pass # try next url_option
        raise ValueError(f"Failed to download {filepath}")
    if type(url) != 'string':
        raise TypeError("Argument 'url' must be of type list or string")
    ...
```

On this solution I still have 2 concerns:
1. The except statement should only filter on certain types of errors, like the 403 error we had before. I suggest for now to only handle the known errors rather than 'any' error (best practise):
```py
except urllib.error.HTTPError as e:
    pass # try next url_option
```
2. When we failed to download from any url, we should not `raise ValueError(...)` as this is not exactly the correct error, but it's the best I could come up with for now...

Please try this yourself and commit this. (I am short on time and I did not test this out yet). If it works, try it out by adding another (invalid) url to the `vulkanInstallURL` and the `VulkanSDKDebugLibsURL`: https://github.com/TheCherno/Hazel/blob/195ed999de4c72316e4fbef28b5244672eb06551/scripts/SetupVulkan.py#L51
https://github.com/TheCherno/Hazel/blob/195ed999de4c72316e4fbef28b5244672eb06551/scripts/SetupVulkan.py#L65

A possible invalid url for you 'should' be displayed in https://github.com/TheCherno/Hazel/pull/438#discussion_r639380014 already, but it seems this is working with this commit? In order to get it to give an error 403, you might have to remove the header (user agent). In this case it should fail to download any url (or that is what I expect at least) and thus raise the ValueError.